### PR TITLE
TF2 for Quantus

### PR DESCRIPTION
**Pull request by dilyabareeva**
- added model interface for Quantus to be compatible with both TF2 and Torch
- unit tests

### CODE REVIEW

**Review by sebastian-lapuschkin — CHANGES_REQUESTED** [review]
let us maybe not make BOTH torch and tensorflow mandatory requirements for the toolkit. The idea was to support both, not force the user to install both. this might, after all, not work out in the end, if the user's model has specific version requirements for  torch/tf/etc.

that being said, it would be good to find a better (long term) solution, which, e.g. via installation instructions, clarifies the optional character of choosing one of both.

additionally, wrt the versioning issue, it might make sense to specifically nail down the (latest) package versions supported, to avoid version clashes. TF and Pytorch are fast moving libraries, and we know from the development of chr5thr/zennit (and others), that it is infeasible to claim (via requirements.txt) that "it just works with the latest version of everything.

**Review by sebastian-lapuschkin — CHANGES_REQUESTED** [review]
While I also think the introduction of a backend-agnostic ModelInterface is the way to go, I think the import of both torch and tf stuff in utils.py is reasonable. as written in the other change request, some users might only want to install (a specific version) of one of the toolkits.

imports and model wrapping should therefore be handled adaptively according to package availability, and warnings or thrown errors should be used to handle if e.g. non of tf or torch are installed, when trying to wrap a backend-native model into a ModelInterface.

python allows the handling of  this "import if possible" idea quite well. some years back, I upgraded my very old numpy-based LRP toolbox to be GPU-capable, if (and only if) the user optionally installs cupy.

see https://github.com/sebastian-lapuschkin/lrp_toolbox/blob/aad9f7d7ebb436d8fe978c34ae78784887e26486/python/modules/linear.py , for example, where cupy is only imported if the user has it installed, which automatically unlocks the optional GPU functionality without the user being bothered with that requirement.

it would be nice to also auto-handle gpu backends similarly, and throw warnings at the user if something goes wrong.

File: quantus/helpers/explanation_func_tf.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    if "method" not in kwargs:
+        warnings.warn(
+            f"Using quantus 'explain' function as an explainer without specifying 'method' (str) "
+            f"in kwargs will produce a vanilla 'Gradient' explanation.\n",
+            category=UserWarning,
+        )
+
+    method = kwargs.get("method", "Gradient").lower()
+
+    inputs = inputs.reshape(
```
Comment (leanderweber) [inline quantus/helpers/explanation_func_tf.py:39]: why not use make use of model.shape_input here?

File: requirements.txt [hunk 2]
```diff
@@ -10,6 +10,8 @@ pytest-lazy-fixture
 scikit-image
 scikit-learn
 scipy
+tensorflow
```
Comment (leanderweber) [inline requirements.txt:13]: again, double tf and pytorch requirements

**Review by leanderweber — CHANGES_REQUESTED** [review]
Made some comments in the respective files.

Also, would be great to have a bit more documentation in the new classes and functions.

File: quantus/helpers/explanation_func_tf.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                )
+            )
+        )
+
+    else:
+        raise KeyError(
+            "Specify a XAI method that already has been implemented {}."
+        ).__format__("XAI_METHODS")
+
+    if isinstance(explanation, torch.Tensor):
```
[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/helpers/explanation_func_tf.py:64]: true @dilyabareeva please see this

File: quantus/helpers/models.py [hunk 4]
```diff
@@ -1,4 +1,6 @@
 import torch
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/helpers/models.py:2]: we resolve this in different PR I would say

File: quantus/helpers/pytorch_model.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+class PyTorchModel(ModelInterface):
+    def __init__(self, model):
+        super().__init__(model)
+
+    def predict(self, x, softmax_act=False, **kwargs):
+        device = kwargs.get("device", None)
+        with torch.no_grad():
+            pred = self.model(torch.Tensor(x).to(device))
+        if softmax_act:
+            return torch.nn.Softmax()(pred).numpy()
```
[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/helpers/pytorch_model.py:16]: @dilyabareeva see this, think this should do the job?

        if explanation.requires_grad:
            explanation = explanation.cpu().detach().numpy()
        else:
            explanation = explanation.cpu().numpy()

File: quantus/metrics/complexity_metrics.py [hunk 6]
```diff
@@ -110,10 +111,14 @@ def __call__(
             >> metric = Sparseness(abs=True, normalise=False)
             >> scores = metric(model=model, x_batch=x_batch, y_batch=y_batch, a_batch=a_batch_saliency, **{}}
         """
+        # Wrap the model into an interface
+        model = get_wrapped_model(model)
+        # Reshape TensorFlow input batch:
```
[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/metrics/complexity_metrics.py:116]: @dilyabareeva

File: quantus/helpers/model_interface.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+from abc import ABC, abstractmethod
+
+
+class ModelInterface(ABC):
+    def __init__(self, model):
```
[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/helpers/model_interface.py:5]: good suggestion

[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/helpers/model_interface.py:5]: @dilyabareeva

File: quantus/helpers/explanation_func_tf.py (hunk 1 above)
Comment (annahedstroem) [reply quantus/helpers/explanation_func_tf.py:39]: @dilyabareeva

Comment (dilyabareeva) [reply quantus/helpers/explanation_func_tf.py:39]: we pass a tf.keras.Model object to this function, not a PyTorchModel, so we can't call this method

**Comment by annahedstroem** [comment]
Hi Dilya, thanks for your work.

A couple of more todos with respect to the explanation_func.py script.

# TODO. Do import of explainability library
# TODO. Figure out model type -> go captum or tf-explain
# TODO. Throw ImportError in case we don't have any explanainability library and print a nice message.
# TODO. Typing of return
# TODO. Add tf-explain methods corresponding to captum (if it does not exist, remove the captum like feture ablation)

thanks!

File: quantus/helpers/explanation_func_tf.py (hunk 1 above)
Comment (leanderweber) [reply quantus/helpers/explanation_func_tf.py:39]: true! however, this restricts explanations to models with channels_last ordering inputs. Since we have the model interface class, I think it makes sense to try and support flexibility where possible (as we should make as few assumptions about the model as possible to improve usability)

So maybe the passed input could be an instance of TensorFlowModel instead of a tf.keras.Model? Or is there a reason not to?

File: quantus/helpers/explanation_func.py [hunk 8]
```diff
@@ -1,22 +1,35 @@
...
+        "Explanation library not found. Please install Captum for torch models and tf-explain for TensorFlow."
+    )
+if util.find_spec("captum"):
+    from captum.attr import *
+if util.find_spec("tf_explain"):
+    import tf_explain
+
+
 def explain(
     model: torch.nn,
```
Comment (leanderweber) [inline quantus/helpers/explanation_func.py:27]: typing of model, inputs and targets parameters is not up-to-date for tensorflow integration, right?

**Review by leanderweber — COMMENTED** [review]
just one comment

**Review by annahedstroem — COMMENTED** [review]
Dilya! thanks for the update on explanation_func.py 

Everything looks good, the only thing missing is docstrings (they do not have to be elaborate but there should be a one-liner about what the function is doing).

Please add at minimum one-liner docstrings to the explanation_func.py script as well as to all other additional functions made in this PR.

**Comment by dilyabareeva** [comment]
There is also another problem with transform_to_normalized_grayscale function in tf-explain. It is basically applied to every explanation in the end and normalises it. So for tf models, where we use tf-explain, we currently can't allow normalise=false argument in explain() function, because tf-explain is going to normalise them anyways. (also it multiplies the result by 255 and outputs integers. sure we can divide it again by 255 in the explain function, but it would be a bit imprecise). See my latest commit where I address this.

File: quantus/helpers/explanation_func.py [hunk 9]
```diff
@@ -1,26 +1,39 @@
...
 import random
+from importlib import util
 import cv2
 from captum.attr import *
 import warnings
 from .utils import *
 from .normalise_func import *
 
 
+if not (util.find_spec("captum") or util.find_spec("tf_explain")):
```
Comment (annahedstroem) [inline quantus/helpers/explanation_func.py:16]: move to def explain

File: quantus/helpers/explanation_func.py [hunk 10]
```diff
@@ -1,26 +1,39 @@
...
+    raise ImportError(
+        "Explanation library not found. Please install Captum for torch models and tf-explain for TensorFlow."
+    )
+if util.find_spec("captum"):
```
Comment (annahedstroem) [inline quantus/helpers/explanation_func.py:20]: make a global flag from this logic

File: quantus/helpers/explanation_func.py (hunk 9 above)
Comment (annahedstroem) [reply quantus/helpers/explanation_func.py:16]: you can sole this by installing quantus by pip install quantus[extras]

File: quantus/helpers/pytorch_model.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
+from copy import deepcopy
+from contextlib import suppress
+
+from ..helpers.model_interface import ModelInterface
+import torch
+import numpy as np
+
+
+class PyTorchModel(ModelInterface):
```
Comment (annahedstroem) [inline quantus/helpers/pytorch_model.py:9]: try out different torch versions (also for tf implementation)

File: quantus/helpers/explanation_func.py (hunk 10 above)
Comment (annahedstroem) [reply quantus/helpers/explanation_func.py:20]: but it in __init__ file

Comment (annahedstroem) [reply quantus/helpers/explanation_func.py:20]: __EXTRAS__ = True if util.find_spec("captum") or util.find_spec("tf_explain")) else False

Comment (sebastian-lapuschkin) [reply quantus/helpers/explanation_func.py:20]: maybe make that `__EXTRAS__ = ...`

Comment (sebastian-lapuschkin) [reply quantus/helpers/explanation_func.py:20]: `__EXTRAS__ = util.find_spec("captum") or util.find_spec("tf_explain")`


but, one question: is the attribution method availability (which surely is not 100% identical between captum and tf_explain covered, or will the user potentially be surprised by  unavailable attribution methods?
